src/main/sensors/compass.c

```diff
@@ -349,6 +355,16 @@ bool compassInit(void)
 
     buildRotationMatrixFromAlignment(&compassConfig()->mag_customAlignment, &magDev.rotationMatrix);
 
+    if (magDev.magOdrHz) {
+        // For Mags that send data at a fixed ODR, we wait some quiet period after a read before checking for new data
+        // allow two re-check intervals, plus a margin for clock variations in mag vs FC
+        uint16_t odrInterval = 1e6 / magDev.magOdrHz;
+        compassReadIntervalUs =  odrInterval - (2 * COMPASS_RECHECK_INTERVAL_US) - (odrInterval / 20);
```

Comment:
Each check is quite expensive. At most single 'data not ready' check shall be performed.

Ideally PLL would be used to monitor MAG update, tuning update frequency so that few percent of checks are too soon. That way, you can stay synchronized to update rate with minimal overhead.

Reply:
Yes, I think that with this code we would, on average, make 1-2 unnecessary checks for each measurement.  If measurement is early, some other measurement is likely to be late.  Tuning the update rate based on measured interval could, I agree, improve matters a little, but not much.